### PR TITLE
Pass ITestHostServices to test runner if it suports it

### PR DESCRIPTION
**Pull request by pakrym**
https://github.com/aspnet/Testing/issues/180
depends on https://github.com/aspnet/dnx/pull/3100
/cc @davidfowl

### CODE REVIEW

**Comment by pakrym** [comment]
Chages in runner: https://github.com/pakrym/aspnet.xunit/commit/f9e9f20aecdc151dd21319268d9365fa6330fe57

File: src/Microsoft.Dnx.Testing.Abstractions/ITestHostServices.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Dnx.Testing.Abstractions
+{
+    public interface ITestHostServices
```
Comment (davidfowl) [inline src/Microsoft.Dnx.Testing.Abstractions/ITestHostServices.cs]: Should probably be a base class or a poco so we can extend it later.

File: src/Microsoft.Dnx.Testing.Abstractions/project.json [hunk 2]
```diff
@@ -8,6 +8,9 @@
   "compilationOptions": {
     "warningsAsErrors": true
   },
+  "dependencies": {
+    "Microsoft.Extensions.Logging": "1.0.0-*"
```
Comment (davidfowl) [inline src/Microsoft.Dnx.Testing.Abstractions/project.json]: This looks wrong. Should be Logging abstractions.

Comment (rynowak) [inline src/Microsoft.Dnx.Testing.Abstractions/project.json]: Nope. Testing boots up a `LoggerFactory`

Comment (davidfowl) [inline src/Microsoft.Dnx.Testing.Abstractions/project.json]: Testing.Abstractions shouldn't be booting anything up.

File: src/Microsoft.Dnx.Testing.Abstractions/ITestHostServices.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.Dnx.Testing.Abstractions/ITestHostServices.cs]: :+1:

Should we just build a static accessor for this thing now a la `PlatformServices`. Really strange for unit testing to have DI when console apps don't.

File: src/Microsoft.Dnx.Testing.Abstractions/project.json (hunk 2 above)
Comment (rynowak) [inline src/Microsoft.Dnx.Testing.Abstractions/project.json]: Oh right :laughing:

**Comment by dougbu** [comment]
> Chages in runner: pakrym/aspnet.xunit@f9e9f20

That repo is a fork. Have you submitted a PR for [xunit/dnx.xunit](https://github.com/xunit/dnx.xunit)?

**Comment by pakrym** [comment]
No, we are too far ahead of it in our fork (https://github.com/aspnet/aspnet.xunit) so i can not do PR (http://i.imgur.com/K19Tetx.gifv )

File: src/Microsoft.Dnx.TestHost/ProjectCommand.cs [hunk 3]
```diff
@@ -8,14 +8,15 @@
 using Microsoft.Dnx.Runtime.CommandParsing;
 using Microsoft.Dnx.Runtime.Common;
 using Microsoft.Extensions.PlatformAbstractions;
+using Microsoft.Dnx.Testing.Abstractions;
```
Comment (dougbu) [inline src/Microsoft.Dnx.TestHost/ProjectCommand.cs]: Interesting new alphabet you've got there.

File: src/Microsoft.Dnx.TestHost/ProjectCommand.cs [hunk 4]
```diff
@@ -62,10 +61,21 @@ private static string GetVariable(IApplicationEnvironment environment, string ke
...
+            object instance;
+            MethodInfo entryPoint;
+            if (!EntryPointExecutor.TryGetEntryPoint(assembly, serviceProvider: null,  instance: out instance,  entryPoint: out entryPoint))
+            {
+                return Task.FromResult(1);
+            }
+            var hostEntryPoint = entryPoint.DeclaringType.GetRuntimeMethods()
+                .FirstOrDefault(m => m.Name == "HostedMain");
+            if (hostEntryPoint != null)
+                return EntryPointExecutor.Execute(instance, hostEntryPoint, new object[] { services, args});
```
Comment (dougbu) [inline src/Microsoft.Dnx.TestHost/ProjectCommand.cs]: Braces

File: src/Microsoft.Dnx.Testing.Abstractions/ITestHostServices.cs (hunk 1 above)
Comment (pakrym) [inline src/Microsoft.Dnx.Testing.Abstractions/ITestHostServices.cs]: What if we would need to run tests in parallel?

**Comment by dougbu** [comment]
> we are too far ahead of it in our fork

Then we have an urgent need to rebase our fork. Should not have let ourselves fall that far behind.

**Comment by pakrym** [comment]
@dougbu We are ahed, not behid.

**Comment by dougbu** [comment]
@pakrym then we have been terribly remiss in submitting changes to the upstream repo. I only see one PR open there and it's more than 3 weeks old. But at least @BrennanConroy did the right thing.

**Comment by dougbu** [comment]
Note: This repo exists so that we can respond to DNX breaking changes quickly. But we need to ensure the runner customers use remains compatible with our major releases. dnx.xunit was Beta 8-compatible when that shipped. Big deal if it wasn't. Either way, now need to catch up w/ RC1.

**Comment by pakrym** [comment]
:up: :date:

**Comment by Eilon** [comment]
FYI I logged https://github.com/aspnet/Universe/issues/316 for @dougbu  to update the xunit fork.

File: src/Microsoft.Dnx.TestHost/ProjectTestHostServices.cs [hunk 5]
```diff
@@ -8,40 +8,35 @@
...
-    internal static class TestServices
+    internal class ProjectTestHostServices: TestHostServices
     {
-        public static ServiceProvider CreateTestServices(
+        public override ITestDiscoverySink TestDiscoverySink { get; }
+        public override ITestExecutionSink TestExecutionSink { get; }
+        public override  ISourceInformationProvider SourceInformationProvider { get; }
+        public override ILoggerFactory LoggerFactory { get; }
+
+        public ProjectTestHostServices(
```
Comment (dougbu) [inline src/Microsoft.Dnx.TestHost/ProjectTestHostServices.cs]: Strange to place the constructor after properties.

File: src/Microsoft.Dnx.Testing.Abstractions/ITestHostServices.cs (hunk 1 above)
Comment (dougbu) [inline src/Microsoft.Dnx.Testing.Abstractions/ITestHostServices.cs]: @pakrym what's the concern? The runner invoked handles parallel testing itself.

File: src/Microsoft.Dnx.TestHost/ProjectCommand.cs [hunk 6]
```diff
@@ -62,10 +61,75 @@ private static string GetVariable(IApplicationEnvironment environment, string ke
...
+            if (entryPoint == null)
+            {
+                System.Console.WriteLine("'{0}' does not contain a 'TestMain' method suitable for an entry point", name);
+                return false;
+            }
+
+            instance = programType.GetTypeInfo().IsAbstract ? null : Activator.CreateInstance(programType);
+            return true;
+        }
+        private static MethodInfo GetEntryPoint(Assembly assembly)
```
Comment (dougbu) [inline src/Microsoft.Dnx.TestHost/ProjectCommand.cs]: Add a blank line here.

File: src/Microsoft.Dnx.TestHost/ProjectTestHostServices.cs [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+using Microsoft.Dnx.Testing.Abstractions;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Dnx.TestHost
+{
+    internal class ProjectTestHostServices: TestHostServices
+    {
+        public ProjectTestHostServices(
+               Project project,
+               ReportingChannel channel)
```
Comment (dougbu) [inline src/Microsoft.Dnx.TestHost/ProjectTestHostServices.cs]: Don't make it harder on the next person who edits this file. Just leave the parameters where VS puts them.

File: src/Microsoft.Dnx.TestHost/ProjectTestHostServices.cs [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+            TestDiscoverySink = new TestDiscoverySink(channel);
+            TestExecutionSink = new TestExecutionSink(channel);
+            LoggerFactory = loggerFactory;
+        }
+
+        public override ITestDiscoverySink TestDiscoverySink { get; }
+        public override ITestExecutionSink TestExecutionSink { get; }
+        public override  ISourceInformationProvider SourceInformationProvider { get; }
+        public override ILoggerFactory LoggerFactory { get; }
```
Comment (dougbu) [inline src/Microsoft.Dnx.TestHost/ProjectTestHostServices.cs]: Unusual spacing in this block. Normally have blank lines between members and only a single space after `override`. Same point about the blank lines in the `abstract` base.

File: src/Microsoft.Dnx.TestHost/ProjectCommand.cs [hunk 9]
```diff
@@ -62,10 +61,76 @@ private static string GetVariable(IApplicationEnvironment environment, string ke
...
+
+            var programType = assembly.GetType("Program") ?? assembly.GetType(name + ".Program");
+
+            if (programType == null)
+            {
+                var programTypeInfo = assembly.DefinedTypes.FirstOrDefault(t => t.Name == "Program");
+
+                if (programTypeInfo == null)
+                {
+                    System.Console.WriteLine("'{0}' does not contain a static 'TestMain' method suitable for an entry point", name);
```
Comment (dougbu) [inline src/Microsoft.Dnx.TestHost/ProjectCommand.cs]: Just realized this message is complaining about `TestMain()` when the code can't even find the `Program` class. Wonky.

**Comment by dougbu** [comment]
:shipit: once aspnet/dnx#3100 goes in. The `ILoggerFactory` bit feels messy but was already in this code and the invoked runner. Looking forward to future changes in this area.

**Comment by pakrym** [comment]
@dougbu won't need, https://github.com/aspnet/Testing/pull/179 i just copied apis

**Comment by dougbu** [comment]
> i just copied apis

:ok:

**Pull request merged**